packages/devextreme/js/core/utils/object.js

```diff
@@ -39,16 +39,29 @@ const orderEach = function(map, func) {
     }
 };
 
-const assignValueToProperty = function(target, property, value, assignByReference) {
+const assignValueToProperty = function(target, property, value, extendComplexObject, assignByReference, shouldCopyUndefined) {
     if(!assignByReference && variableWrapper.isWrapped(target[property])) {
         variableWrapper.assign(target[property], value);
+    } else if(!assignByReference && Array.isArray(value)) {
+        target[property] = value.map(item => {
+            let itemTarget = item;
+
+            if(Array.isArray(item)) {
+                itemTarget = [];
+            }
+            if(isObject(item)) {
+                itemTarget = {};
+            }
```

Comment:
Let's move out this logic to a separate function

Something like:
```javascript
const getDeepCopyTarget = (item) => {
   if (isObject(item)) {
     return Array.isArray(item) ? [] : {}
   }
   
   return item;
}
```
And use it in ```map```:
```javascript
target[property] = value.map((item) => {
  const itemTarget = getDeepCopyTarget(item);
  return deepExtendArraySafe(itemTarget, item, extendComplexObject, assignByReference, shouldCopyUndefined);
});
```